test/benchmarks/tools/hackbench.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2022 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tools
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"testing"
+)
+
+// Hackbench makes 'hackbench' commands and parses their output.
+type Hackbench struct {
+	IpcMode     string // ipc mode: pipe, socket(default)
+	ProcessMode string // process mode: thread, process(default)
+}
+
+// MakeCmd makes commands for Hackbench.
+func (s *Hackbench) MakeCmd(b *testing.B) []string {
+	cmd := []string{"hackbench"}
+	// ipc mode
+	if s.IpcMode == "pipe" {
+		cmd = append(cmd, fmt.Sprintf("--pipe"))
+	}
+	// group num
+	cmd = append(cmd, fmt.Sprintf("--groups=10"))
+	// process mode
+	if s.ProcessMode == "thread" {
+		cmd = append(cmd, fmt.Sprintf("--threads"))
+	} else {
+		cmd = append(cmd, fmt.Sprintf("--process"))
+	}
+	// loops
+	cmd = append(cmd, fmt.Sprintf("--loops=1000"))
```

Comment:
I see. But the problem is hackbench uses the execution time as performance metrics. So, if we use b.N, different platform will have different loops, which makes their execution time incomparable. I think maybe we can use `BENCHMARKS_OPTIONS=--test.benchtime=1x` in pipeline script?

The output of ptrace platform with test.benchtime=1x:
```
goos: linux                                                                                                                                                                 goarch: amd64
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkHackbench                                                                                                                                                          
BenchmarkHackbench/ipcMode.pipe/processMode.thread
BenchmarkHackbench/ipcMode.pipe/processMode.thread-16                  1        23778641994 ns/op               22.61 execution_time.s                                      
BenchmarkHackbench/ipcMode.socket/processMode.process                                                                                                                       
BenchmarkHackbench/ipcMode.socket/processMode.process-16               1        157532874461 ns/op             155.4 execution_time.s
```

The output of runc platform with test.benchtime=1x:
```
goos: linux
goarch: amd64
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkHackbench
BenchmarkHackbench/ipcMode.pipe/processMode.thread
BenchmarkHackbench/ipcMode.pipe/processMode.thread-16                  1        1239167610 ns/op                 0.6450 execution_time.s                                    
BenchmarkHackbench/ipcMode.socket/processMode.process
BenchmarkHackbench/ipcMode.socket/processMode.process-16               1        1066953808 ns/op                 0.5520 execution_time.s
```

Reply:
We have other benchmarks that have this property...mainly ones we really only want to run once. Golang's `test.benchtime=1x` [does not always behave the way you'd think](https://github.com/golang/go/issues/32051), so we have an override hack in the afformatentioned benchmarks, we force `--test.benchtime=1ns`, which is .

https://github.com/google/gvisor/blob/cfc29d3b5dacc8b18acf69b39919afc1756820c5/test/benchmarks/fs/bazel_test.go#L59

We also have benchmarks where we flat out ignore the ns/op metric because there is too much noise or it doesn't matter. But we've still tried to stick to the standard of still scaling by b.N.

I suggest the "1ns" hack for this one.

Reply:
ok, updated